### ISSUE TITLE
ENH refactor Adversarial Debiasing

### ISSUE THREAD

**Issue by SeanMcCarren** [fairlearn/fairlearn#1082]
Should the following be an issue or a comment in the code?

In the future, we may want to expose the AdversarialFairness base class to allow for arbitrary loss functions and multi-output mix-type targets/sensitive features (although the latter doesn't seem very important). But to do that, some refactoring is required:
- Do not handle/predict all preprocessing, but instead be reactive to the data. That way we don't have to predict the shape of the (partly) one-hot-encoded matrices. So, also create neural nets after preprocessing instead of before.
- try to get rid of _preprocessor.py, and force users to specify loss functions in weird cases. The type_of_target style inference in _preprocess.py fits just as well in the base class.
